Source/JSON/JSONFactory.swift

```diff
@@ -0,0 +1,49 @@
+//
+//  JSONFactory.swift
+//  Matrioska
+//
+//  Created by Mathias Aichinger on 11/01/2017.
+//  Copyright © 2017 runtastic. All rights reserved.
+//
+
+import Foundation
+
+/// An error type for JSONFactory
+///
+/// - missing: specifies that a mandatory key is missing
+enum JSONFactoryError: Error {
+    case missing(JSONObject, String)
+}
+
+/// A factory that wraps ComponentFactory objects and uses them to produce Components
+public final class JSONFactory {
+    
+    let typeKey = "type"
+    let metaKey = "meta"
+    let childrenKey = "children"
+    
+    /// The already registered factories
+    private(set) var factories: [String: ComponentFactory] = [:]
+    
+    /// Registers a new ComponentFactory which will be used in `produce`
+    ///
+    /// - Parameter factory: the ComponentFactory to be registered
+    public func register(_ factory: ComponentFactory) {
+        factories[factory.typeName()] = factory
+    }
+    
+    /// Produces a Component from a given JSONObject, which has one mandatory key: `typeKey`
+    ///
+    /// - Parameter json: the JSONObject to be used
+    /// - Returns: An optional Component
+    /// - Throws: JSONFactoryError when a mandatory key is missing
+    public func produce(from json: JSONObject) throws -> Component? {
+        guard let type = json[typeKey] as? String else { throw JSONFactoryError.missing(json, typeKey) }
+        
+        let meta = json[metaKey] as? [String : Any]
+        let children = json[childrenKey] as? [JSONObject] ?? []
+        let componentChildren = try children.flatMap { try produce(from: $0) }
+        
+        return factories[type]?.produce(children: componentChildren, meta: meta)
```

Comment:
I like this approach now, because it's using closures everywhere. The first approach was a little bit a mixture of the closure and class approach. I still prefer the class approach, but probably because I'm thinking in old concepts ;) (I would have used classes for the clusters and protocols for the factories.)